### PR TITLE
Rustup

### PR DESCRIPTION
**Pull request by marcusklaas**
Hi Alex,

This should make alll the code and tests should compile again. About half the tests are still failing, but I thought I'd go ahead and make a pull request anyway. Maybe we can use this PR to discuss what needs to change to get the remaining tests to pass again.

Best,
Marcus

### CODE REVIEW

File: src/bwt/mtf.rs [hunk 1]
```diff
@@ -148,14 +153,14 @@ impl<R> Decoder<R> {
...
 }
 
-impl<R: Reader> Reader for Decoder<R> {
-    fn read(&mut self, dst: &mut [u8]) -> old_io::IoResult<usize> {
+impl<R: Read> Read for Decoder<R> {
+    fn read(&mut self, dst: &mut [u8]) -> io::Result<usize> {
         let mut bytes_read = 0;
         for sym in dst.iter_mut() {
             let rank = match self.r.read_u8() {
+                Ok(0) => break,
```
Comment (kvark) [inline src/bwt/mtf.rs]: shouldn't here be `if bytes_read!=0` condition?

File: src/entropy/ari/test.rs [hunk 2]
```diff
@@ -209,14 +210,14 @@ fn roundtrips_apm() {
...
-    let mut storage: Vec<u8> = repeat(0u8).take(TEXT_INPUT.len()).collect();
-    bh.iter(|| {
-        let mut w = BufWriter::new(storage.as_mut_slice());
-        w.seek(0, SeekSet).unwrap();
-        let mut e = super::ByteEncoder::new(w);
-        e.write(TEXT_INPUT).unwrap();
-    });
-    bh.bytes = TEXT_INPUT.len() as u64;
-}
+// #[bench]
```
Comment (kvark) [inline src/entropy/ari/test.rs]: Benches are no more in Rust?

File: src/flate.rs [hunk 3]
```diff
@@ -5,12 +5,14 @@
 //!
 //! # Example
 //!
-//! ```rust
+//! ```rust, ignore
```
Comment (kvark) [inline src/flate.rs]: what does ignore do here?

File: src/bwt/mtf.rs (hunk 1 above)
Comment (marcusklaas) [inline src/bwt/mtf.rs]: I think you are correct. I will look into it.

File: src/entropy/ari/test.rs (hunk 2 above)
Comment (marcusklaas) [inline src/entropy/ari/test.rs]: Woops -- I don't know why I commented it out.

File: src/flate.rs [hunk 4]
```diff
@@ -209,25 +213,25 @@ impl<R: Reader> Decoder<R> {
...
-            slice::bytes::copy_memory(self.output.slice_from_mut(self.outpos),
-                                    self.block.slice(from, from + n));
+            slice::bytes::copy_memory(&self.block[from..(from + n)],
+                                      &mut self.output[self.outpos..]);
         }
         self.outpos += n;
         if n < amt {
-            slice::bytes::copy_memory(self.output.as_mut_slice(),
-                                    self.block.slice_from(from + n));
+            slice::bytes::copy_memory(&self.block[from..(from + n)],
```
Comment (kvark) [inline src/flate.rs]: This is incorrect. Should be `self.block[(from+n)..]`

File: src/flate.rs (hunk 3 above)
Comment (marcusklaas) [inline src/flate.rs]: It makes cargo test ignore this example as a test. It would otherwise fail because of the file access.

File: src/flate.rs [hunk 5]
```diff
@@ -454,15 +458,15 @@ impl<R: Reader> Decoder<R> {
...
         if self.pos == self.block.len() {
-            if self.eof { return Err(old_io::standard_error(old_io::EndOfFile)) }
+            if self.eof { return Ok(0) }
             try!(self.block());
         }
         let n = cmp::min(buf.len(), self.block.len() - self.pos);
-        slice::bytes::copy_memory(buf.slice_to_mut(n),
-                                self.block.slice(self.pos, self.pos + n));
+        slice::bytes::copy_memory(&buf[..n],
+                                  &mut self.block[self.pos..(self.pos + n)]);
```
Comment (kvark) [inline src/flate.rs]: This whole invocation is incorrect. Should be:
```rust
slice::bytes::copy_memory(&self.block[self.pos..(self.pos + n)],
                          &mut buf[..n]);
```

File: src/flate.rs [hunk 6]
```diff
@@ -518,13 +526,24 @@ mod test {
...
             }
         }
+
+        let expected = include_bytes!("data/test.txt");
+
         assert!(d.eof());
-        assert!(out.as_slice() == include_bytes!("data/test.txt"));
+        assert_eq!(expected.len(), out.len());
+
+        for (i, (a, b)) in out.iter().zip(expected.iter()).enumerate() {
```
Comment (kvark) [inline src/flate.rs]: why is this loop needed?

File: src/lz4.rs [hunk 7]
```diff
@@ -524,19 +531,24 @@ impl<W: Writer> Encoder<W> {
...
-    pub fn finish(mut self) -> (W, old_io::IoResult<()>) {
-        let result = self.flush();
-        let result = result.and(self.w.write_le_u32(0));
-        // XXX: this checksum is wrong
-        let result = result.and(self.w.write_le_u32(0));
+    pub fn finish(mut self) -> (W, io::Result<()>) {
+        let mut result = self.flush();
+
+        for _ in 0..2 {
+            let tmp = self.w.write_u32::<LittleEndian>(0)
```
Comment (kvark) [inline src/lz4.rs]: nit pic: could use a `fold` here

**Comment by kvark** [comment]
@marcusklaas Thank you for the enormous amount of work in updating the source! Very much appreciated :) I've finished the review, will gladly merge when you fix my findings.

If tests are still going to be failing, that means we either missed something, or there are more semantic changes outside of `copy_bytes`. @alexcrichton would know better.

File: src/flate.rs (hunk 4 above)
Comment (marcusklaas) [inline src/flate.rs]: Good catch!

File: src/flate.rs (hunk 6 above)
Comment (marcusklaas) [inline src/flate.rs]: It was added for debugging purposes, forgot to remove it.

**Comment by marcusklaas** [comment]
Thank you kindly for your swift and thorough feedback @kvark! I will get to work on those issues immediately.

File: src/bwt/mod.rs [hunk 8]
```diff
@@ -351,42 +357,41 @@ impl<R: Reader> Decoder<R> {
...
 
-    fn decode_block(&mut self) -> old_io::IoResult<bool> {
-        let n = match self.r.read_le_u32() {
+    fn decode_block(&mut self) -> io::Result<bool> {
+        let n = match self.r.read_u32::<LittleEndian>() {
+            Ok(0) => return Ok(false),
             Ok(n) => n as usize,
-            Err(ref e) if e.kind == old_io::EndOfFile => return Ok(false),
-            Err(e) => return Err(e),
+            Err(e) => return Err(byteorder_err_to_io(e))
```
Comment (marcusklaas) [inline src/bwt/mod.rs]: This was also a source of error. It no longer returned `Ok(false)` on EOF.

Comment (kvark) [inline src/bwt/mod.rs]: nice that you spotted it!

**Comment by marcusklaas** [comment]
The remaining tests should now pass as well, thanks for the pointers @kvark!

File: src/lz4.rs (hunk 7 above)
Comment (marcusklaas) [inline src/lz4.rs]: This is the only point I didn't address. I don't see how a fold would work here.. Could you give me a pointer?

Comment (kvark) [inline src/lz4.rs]: It's really minor, we can do it any time later:
```rust
let result = (0..2).fold(self.flush(), |result, _| {
    let tmp = self.w.write_u32::<LittleEndian>(0);
                  .map_err(byteorder_err_to_io);
    result.and_then(|_| tmp)
});

**Comment by kvark** [comment]
Wuut, tests are passing, that's extra sweet!
Thanks again, @marcusklaas !

**Pull request merged**